pkg/gatewayserver/gatewayserver_test.go

```diff
@@ -556,11 +556,18 @@ func TestGatewayServer(t *testing.T) {
 					},
 				} {
 					t.Run(ctc.Name, func(t *testing.T) {
-						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(timeout))
-						upCh := make(chan *ttnpb.GatewayUp)
-						downCh := make(chan *ttnpb.GatewayDown)
-						err := ptc.Link(ctx, t, ctc.ID, ctc.Key, upCh, downCh)
-						cancel()
+						var err error
+						for i := 0; i < 5; i++ {
```

Comment:
I don't fully understand this. Are we trying 5 times with the same timeout to link? And hope it works one of those 5 times?

Reply:
Yes, see comment below

Reply:
This is not acceptable

Reply:
We repeat only if the error is that the context expired early, for a max of 5 times. Maybe it is better to generate a proper event and check for that on the test.

Reply:
See above

An event is also a good solution

Reply:
Changed test to use the `gs.gateway.connect` event. Can you make sure all cases are checked as before (also cc @KrishnaIyer)